### DISCUSSION TITLE
[Preview Discussion] Argument Coercion Proposal

### DISCUSSION THREAD

**Discussion by TotalTechGeek** [json-logic/.github#2] [General]
Hey folks 👋🏻 

We still need to set community guidelines and set up a process and repository for presenting proposals.

I wanted to bring up one item a bit early that I feel might be controversial and may not end up being adopted. I fully expect to lose, but I hope that I'll be able to convice 😉 

### Background

So, to get things started, let's establish some background, JSON Logic is basically an AST / serializable S-Expressions in JSON. 

To call a method, you use a key in an object, and the value is treated as your arguments for the expression.

```json
{ "max": [1, 2, 3] }
```
 
 Is conceptually the same as calling `max(1, 2, 3)`
 
So far so good. 
 
Per mainline JSON Logic, when you call a method directly a value, this is treated as shorthand for calling with a single argument array.

Therefore, `{ "max": 1 }` is the same as `{ "max": [1] }`, which translates to `max(1)`

This is also fairly reasonable, though has caused some confusion with `{ "not": 0 }` vs `{ "not": [0] }`. Both of these expressions should translate to `not(0)` because of the shorthand sugaring; but I ultimately believe this design decision makes sense, and I believe needs to be preserved.

### Proposal

Where I diverge slightly is in how I wish to treat chained logic statements.

Imagine I have `{"max": {"var": "data"} }` and `{ "data": [1, 2, 3] }`

If we treat chained logic the same way as values, we limit our capabilities significantly, *or* make operator implementation quite a bit more complex.

For example, if `{"max": {"var": "data"} }` is interpreted as `{"max": [{"var": "data"}] }`, it would be equivalent to `max([1, 2, 3])` instead of `max(1, 2, 3)`, which unless we create a rather robust implementation of `max`, would likely return awkward results.

Similarly, if I had `{ "cat": { "merge": [["Hello "], ["World", "!"]] } }`, cat would receive `[["Hello ", "World", "!"]]` as its argument list. 

In `json-logic-js`, this ends up producing `"Hello ,World,!"` as it automatically calls `toString` on the array..

What if Instead of the sugaring operation mentally being `(x is not array) ? [apply x] : apply x`, the sugaring operation could instead be `concat([], apply x)`.

Then we'd be able to chain function calls and essentially treat them as if they spread into the argument list, so if I say:
```json
{ "cat": { "var": "text" } }
```

With the data
```json
{ "text": ["Hello ", "World", "!"] }
```

It will result in `Hello World!`

So instead of `{ "cat": { "var": "text" } }` being automatically sugared into `{ "cat": [{ "var": "text" }] }` as `cat([var("text")])`, it's instead parsed as `cat(...var("text"))`

Making it possible to do things like:

```json
{
  "max": {
      "map": [
        {
          "filter": [
            { "var": "people" }, 
            { "===": [{ "var": "department" }, "Engineering"] }
          ]
        }, 
        { "var": "salary" }
    ]
  }
}
```

With the following data:
```json
{
  "people": [
    { "name": "Jay Ortiz", "salary": 100414, "department": "Engineering"},
    { "name": "Louisa Hall", "salary": 133601, "department": "Sales"},
    { "name": "Kyle Carlson", "salary": 139803, "department": "Sales"},
    { "name": "Grace Ortiz", "salary": 147068, "department": "Engineering"},
    { "name": "Isabelle Harrington", "salary": 112704, "department": "Marketing"},
    { "name": "Harold Moore", "salary": 125221, "department": "Sales"},
    { "name": "Clarence Schultz", "salary": 127985, "department": "Sales"},
    { "name": "Jesse Keller", "salary": 149212, "department": "Engineering"},
    { "name": "Phillip Holland", "salary": 105888, "department": "Marketing"},
    { "name": "Mason Sullivan", "salary": 147161, "department": "Engineering" }
  ]
}
```

And output `149212`

### Arguments Against

I believe a solid argument against this proposal is that it diverges significantly from the current mental model for the sugaring operation.

`concat([], apply x)` is a bit more advanced of an operation to deal with sugaring.

### Arguments For

This approach would make JSON Logic significantly more flexible and capable. 

It'd be easy to pass the output of a `filter` & `map` statement into `cat` or `+` and have it produce a reasonable result. 

It also provides this capability without embedding a ton of edge cases into each operator to be able to support "If I receive a single argument, and that argument is a list"

It's also still the same operation, it just changes when the "is array" evaluation occurs by one step.

### Backwards Compatibility 

I don't believe this tweak would break anyone's existing logic; and I believe that this would enable use-cases that otherwise would not be possible.

**Comment by TotalTechGeek** [comment]
Note: Going forward, I think that proposals should be presented in JSON Blocks,

```json
[
  "Basic Tests",
  {
    "description": "Adds two numbers together",
    "rule": { "+": [1, 2] }
    "result": 3,
    "data": {}
  }
]
```

Format outlined here: https://github.com/json-logic/.github/blob/main/TEST_FORMAT.md

To make it easier to extract and play with. 

Here are some of the proposed behaviors:


```json
[
    "# These are tests from https://github.com/orgs/json-logic/discussions/2",
    {
        "description": "Standard Max",
        "rule": { "max": [1, 2, 3] },
        "data": {},
        "result": 3
    },
    {
        "description": "Standard Max, Single Argument Sugared",
        "rule": {"max": 1},
        "data": {},
        "result": 1
    },
    {
        "description": "Max with Logic Chaining",
        "rule": { "max": {"var": "data"} },
        "data": { "data": [1, 2, 3] },
        "result": 3
    },
    {
        "description": "Cat with Logic Chaining",
        "rule": {
            "cat": {
                "merge": [ ["Hello "], ["World", "!"] ]
            }
        },
        "data": {},
        "result": "Hello World!"
    },
    {
        "description": "Cat with Logic Chaining (Simple)",
        "rule": { "cat": {"var": "text"} },
        "data": { "text": ["Hello ", "World", "!"] },
        "result": "Hello World!"
    },
    {
        "rule": {
            "max": {
                "map": [
                    {
                        "filter": [
                            {"var": "people"},
                            {
                                "===": [ {"var": "department"}, "Engineering" ]
                            }
                        ]
                    },
                    {"var": "salary"}
                ]
            }
        },
        "data": {
            "people": [
                {"name": "Jay Ortiz"          , "salary": 100414, "department": "Engineering"},
                {"name": "Louisa Hall"        , "salary": 133601, "department": "Sales"      },
                {"name": "Kyle Carlson"       , "salary": 139803, "department": "Sales"      },
                {"name": "Grace Ortiz"        , "salary": 147068, "department": "Engineering"},
                {"name": "Isabelle Harrington", "salary": 112704, "department": "Marketing"  },
                {"name": "Harold Moore"       , "salary": 125221, "department": "Sales"      },
                {"name": "Clarence Schultz"   , "salary": 127985, "department": "Sales"      },
                {"name": "Jesse Keller"       , "salary": 149212, "department": "Engineering"},
                {"name": "Phillip Holland"    , "salary": 105888, "department": "Marketing"  },
                {"name": "Mason Sullivan"     , "salary": 147161, "department": "Engineering"}
            ]
        },
        "result": 149212,
        "description": "Max with Logic Chaining (Complex)"
    }
]
```

**Comment by TotalTechGeek** [comment]
@json-logic/core Hey folks, this is a "test" proposal to solicit feedback on the best way to organize proposals. 

I do think that each proposal should have the relevant JSON attached.

If y'all have suggestions on how to organize proposals going forward, we can definitely incorporate that into the initiative's guidelines

**Reply by codetiger** [reply]
If you look at the original test cases it follows
```
[
  "Optional Title",
  [
    /* Rule */,
    /* Data */,
    /* Result */
  ],
]
```

I would suggest keeping it compatible with test case format, as it has a small advantage of reusing it as test data. So my proposal: 
```
[
  "Optional Title",
  [
    /* Rule */,
    /* Data */,
    /* Result */,
    /* Optional Description */
  ],
]
```

**Reply by TotalTechGeek** [reply]
100% Agree!

**Reply by gregsdennis** [reply]
I would actually take this opportunity to redo the format as something a bit more explicit. Personally, I find positional definitions hard to read, especially as the data gets larger, which will likely be the case for this project.

Please have a look at the JSON Schema test suite for an example. (Not saying that's a perfect example; there is plenty of room for improvement.)

**Reply by TotalTechGeek** [reply]
Are you thinking something like

```typescript
{
  "description": "Adds two numbers together",
  "rule": { "+": [1, 2] }
  "result": 3,
  "data": {}
}
```

Or more along the lines of:

```typescript
{
  "description": "Adds two numbers together",
  "rule": { "+": [1, {"var": "a"}] }
  "tests": [
    { "data": { "a": 1 }, "result": 2, "description": "Positive number"  },
    { "data": { "a": -1 }, "result": 0, "description": "Negative number" }
   ]
}
```

**Reply by gregsdennis** [reply]
Either of those would be fine. They're explicit and easy to read. I know exactly what each part does.

**Reply by TotalTechGeek** [reply]
In that case how about we opt for the first guy keeps it simple to traverse and report on. 

@codetiger Is that alright?

**Reply by TotalTechGeek** [reply]
Hey folks, I've introduced a simple definition here, https://github.com/json-logic/.github/blob/main/TEST_FORMAT.md

This can be formalized a bit more, this is just to get the convention documented.

**Reply by josephdpurcell** [reply]
Some quick thoughts and I'm still getting up to speed so I could be missing things:

The test structure should be well typed to avoid type bugs between languages that could be avoided.

The grouping concept feels iffy? I would like to see something like:

```
{ "groupId": [] }
```

I do like the idea of groups but would like to see a simpler type structure.

The description as string and required is excellent. I do imagine lots of bike shedding over what an appropriate description is. Not saying that's bad or good just observing. https://www.conventionalcommits.org/en/v1.0.0/#specification comes to mind, an entire spec just on a tiny bit of text (albeit VERY important text).

The rule as an any type is problematic. I would expect the rule type to be rigorously defined in its structure. This rigor applied to the rule will keep engine/parsing to a minimal. How should `true` as a rule be processed? An easy improvement would be to say it's an object. In my time musing about rules I think the first-level of a rule should be an object and if that rule wants to be some combination of other rules then there should be a "combination" operator. I feel most strongly about this opinion, that rule should be rigorously typed, compared to the other thoughts here. But this could be an area where I'm missing some pieces about what a JSON Logic rule is.

I think I object to data being an object? Couldn't data be `true` or any type? There have been cases where I wanted to do logic against a non-object. And as I'm writing this I think I feel as strongly if not more so about this opinion. Additionally I don't think it should default to an empty object. The field should be required and if it's omitted the expectation is the engine should treat it as the equivalent of "undefined" in that language. The test should explicitly say what it's testing and it doesn't seem fitting to require all engines to implement the very tricky logic of saying "undefined means default to empty object". Imagine you want to apply JSON Logic to user input that is missing? By applying it against an empty object is over-specifying how the engine should handle the case of undefined.

Result as an any and required makes sense but it needs to be clearer. I presume any means any type that JSON RFC 8259. But that means the result cannot be undefined. So does this mean engines must return null if the result is "missing (aka undefined)"? I would like to see the result be any JSON type or the equivalent of "undefined" in whatever language the engine is implemented in.



There's an opportunity for a field that is an array of strings where each string is an identifier of a spec. This would allow mapping tests to specs and vice versa. This would assume specs (or proposals if that's their name?) have an identifier.

**Reply by josephdpurcell** [reply]
One more thing popped up, what is the expectation for a "passing" test? I understand it to be that a given engine takes the data and runs the rule to get the result, and then a deep and strict equivalence is done to match the test result to the engines result. But can we conceive of cases that languages will have trouble with this? Or cases where the result is non-deterministic? I can't think of any right now but it's worth validating.

Or maybe this is a topic that is addressed by saying there will be test cases the engine must implement due to language-specific challenges that are not appropriate to be made into general specification. For example, do all languages even have an undefined type?

My conclusion having explored this for 60 seconds is to handle it similar to PHP FIG where the test cases are more like examples. "Here are examples of JSON Logic behavior that should be consistent across engines, anything not specified is up to the engine to provide the rigorous testing/definition."

**Reply by TotalTechGeek** [reply]
> The grouping concept feels iffy?

There is no grouping concept; it's just inherited comments from `https://jsonlogic.com/tests.json`. 

I should probably tweak the markdown to make it clear that it's just a way to embed some additional human-readable data into it. 

It's not intended to be used by the test suite.

> The rule as an any type is problematic. I would expect the rule type to be rigorously defined in its structure

The rule has to be an any type. I could potentially re-describe it as "any valid JSON value" (JSON RFC 8259), but mainline JSON Logic says that `true` is a valid rule that is evaluated to be `true`. I do not want to break backwards compatibility / redefine that fundamental. 

It's also a detail that most engines use to evaluate arguments, so that `logic.apply` can be run recursively.

> I think I object to data being an object

This is a documentation mistake. Looking at the file will show that this is not actually a requirement.

You are correct, you might pass in `5` or `[1, 2, 3]` as data 😄 

> Additionally I don't think it should default to an empty object. The field should be required and if it's omitted the expectation is the engine should treat it as the equivalent of "undefined" in that language.

`null` would be more appropriate, given that `undefined` does not exist in JSON. 

I think this might be bikeshed territory, and this isn't somewhere I particularly feel very strongly one way or another, 

So I think I might make it mandatory as well. Less room for disagreement 😆

**Reply by TotalTechGeek** [reply]
I've made some tweaks to the doc.

>  Or cases where the result is non-deterministic?

Non-deterministic results would be rather hard to test for 😉. 

But I get what you're referring to; 

For example, if we proposed an extension that allowed someone to use the current time, then the result would naturally be non-deterministic and hard to encode into the spec.

I think this serves as a relatively decent v1 that can handle 99% of our use-cases.

I think we can cover those edge cases and address that if we encounter them; but I don't believe we will be incorporating non-deterministic operators into our specification in the near-term.

I think if we encounter something, we should be able to extend the test specification.

**Reply by TotalTechGeek** [reply]
> then a deep and strict equivalence is done to match the test result to the engines result. But can we conceive of cases that languages will have trouble with this
> equivalent of "undefined" in whatever language the engine is implemented in

I think we'll allow implementers to define equivalence classes in their implementation of the test runner, particularly for that `null` case. 

If the implementer wants to return their language's version of `null` rather than some weird `JSONNull` object, if it would be ergonomic for their users, I think that _should be_ fine.

My reaction to `undefined` was partially driven by none of the currently defined operators being allowed to return `undefined`. `{ var: 'name' }` where the data is `{}` would return `null`.

**Reply by TotalTechGeek** [reply]
> Or cases where the result is non-deterministic?

@beeme1mr Brought up Cucumber in the compat table discussion, and I believe if we encounter more significant scenarios that require non-deterministic outputs (like a spec that uses the current time), Cucumber could help provide the scaffolding necessary to validate these use cases.

I think it tends to be overkill for unit testing, but if we need to measure side-effects or non-deterministic results, it can be an ideal framework for that.

**Comment by gregsdennis** [comment]
I think part of the confusion is that you're trying to solve two problems simultaneously.  Separate the chaining from the "sugaring adjustment".

```json
{ "cat": { "var": "text" } }
```

I propose that there be a two step process. 

1. Resolve all `var`s, and do so iteratively until there are no more. 
2. Process the result as is.

This means that `{ "var": "text" }` in the above would be replaced with the data value for `text`.  If that's an array, e.g. `["x", "y"]`, then process `cat("x", "y")`.  If it's a single value, you first apply the sugaring and then process.

The obvious gotcha is how to retain an array in the data as a single value.  The answer is to require that it explicitly contained in an array.

```json
{ "cat": [{ "var": "text" }] }
```

This kind of thing is done all the time in other syntaxes and languages, where the shorthand can only be used in specific cases, and for other scenarios the full syntax MUST be used.  For example, in RFC 9535 (JSON Path), the dot -syntax for a property selector (`.foo`) can only be used if the property name is alphanumeric (plus underscore) and doesn't start with a number.  Otherwise, you must use the bracketed syntax (`['$foo']`).

It's perfectly acceptable to have cases where the sugaring can't be used as long as we can specify it well, and I think we can in this case.

**Reply by TotalTechGeek** [reply]
I think it's possible that we might be proposing the same thing. Though I want to be explicit about not just resolving `var`, since I'm able to pass `map` into `max`

Given `{ a: x }`

Right now, the common approach is `if x is not an array ? [apply x] : apply x`, where the sugaring happens before the execution.

The model I'm recommending is tweaking it to `concat([], apply x)`, where the array-ification happens *after* the execution like you're suggesting here, when passing the input into another method. 
> Resolve all vars, and do so iteratively until there are no more.

Except my recommendation is "Resolve the logic, and do so recursively / iteratively until there is no more"

Also, this mental model can conceptually remove the concept of sugaring all together, since it could conceptually be executed on *all* traversed arguments of the logic.

- `concat([], apply 1)` would produce `[1]`
- `concat([], apply [1, 2])` would produce `[1, 2]`
- `concat([], apply var('x'))`where `x` is [1,2] would produce `[1,2]`
- `concat([], apply [var('x')])`where `x` is [1,2] would produce `[[1,2]]`

Now, in my implementation, I don't actually implement it using `concat` since it's not as efficient. 

Essentially, it's the same as before, just moving it one step, `y = apply x, if y is not an array ? [y] : y` when passing arguments.

> The obvious gotcha is how to retain an array in the data as a single value. The answer is to require that it explicitly contained in an array.

This is the goal! 

--- 

**A Note** - This would only be executed on arguments for `traversable` operations.

Almost all operations are traversable, a few that aren't: `map`, `some`, `filter`, `if`, etc.

These guys handle their own traversal, so I don't believe we should touch their arguments at all, for "sugaring" or otherwise.

Essentially, if an operator is flagged as `traverse: false`, it should receive the arguments as they're given in the JSON.

**Reply by TotalTechGeek** [reply]
Here's an additional rule this type of thing allows for 

```json
[
  {
    "if": [
      { ">": [{ "max": { "map": [{ "var": "people" }, { "var": "salary" }] } }, 149000 ] },
      "The max salary is above 149K",
      { "<": [{ "min": { "map": [{ "var": "people" }, { "var": "salary" }] } }, 5000 ] },
      "Ok.. the min salary is below 5K",
      "All salaries are between 5-149K"
    ]
  },
  {
    "people": [
      { "name": "Jay Ortiz", "salary": 100414, "department": "Engineering"},
      { "name": "Louisa Hall", "salary": 133601, "department": "Sales"},
      { "name": "Kyle Carlson", "salary": 139803, "department": "Sales"},
      { "name": "Grace Ortiz", "salary": 147068, "department": "Engineering"},
      { "name": "Isabelle Harrington", "salary": 112704, "department": "Marketing"},
      { "name": "Harold Moore", "salary": 125221, "department": "Sales"},
      { "name": "Clarence Schultz", "salary": 127985, "department": "Sales"},
      { "name": "Jesse Keller", "salary": 149212, "department": "Engineering"},
      { "name": "Phillip Holland", "salary": 105888, "department": "Marketing"},
      { "name": "Mason Sullivan", "salary": 147161, "department": "Engineering" }
    ]
  },
  "The max salary is above 149K",
  "Showing how operations can be chained to max, which is then used to check for ranges"
]
```  

Which can be tinkered with here: https://jessemitchell.me/json-logic-engine/

**Reply by gregsdennis** [reply]
> Though I want to be explicit about not just resolving var, since I'm able to pass map into max

So the process would be resolved bottom-up, that is, from leaf to root.

If at any point a non-array is returned where an array is required, it is "wrapped" in an array and processed accordingly.  If the logic author wants an array to be considered as a value, they MUST explicitly include the "parameter" array.

> [!NOTE]
> My implementation doesn't perform this wrapping and instead merely operates directly on the value.  I'm not sure the site [strictly speaking, it's not a spec] explicitly declares that a raw value is shorthand for a parameter array; just that the system also works with raw values.  It does say "we support syntactic sugar on unary operators to skip the array around values", but that's hardly instruction to an implementation that raw values should be treated as param arrays.  If this is the direction we want to take this, which could make sense, this operation needs to be more explicit in a specification.

> if an operator is flagged as `traverse: false`

We need to be careful about specifying this in a language/implementation-agnostic way.  This statement reads as influenced by a specific implementation since the site says nothing about the programmatic structure of an operator (and the spec shouldn't say such things).  The specification should merely indicate operational requirements and these kinds of details are part of how an implementation decided how it wants to meet those requirements.

(This is a learned art.)

**Reply by gregsdennis** [reply]
> `concat([], apply 1)`

This syntax doesn't make sense to me.  I would avoid using language (even pseudo-language) syntax.

**Reply by TotalTechGeek** [reply]
> So the process would be resolved bottom-up, that is, from leaf to root.
>
> If at any point a non-array is returned where an array is required, it is "wrapped" in an array and processed accordingly. If the logic author wants an array to be considered as a value, they MUST explicitly include the "parameter" array.

👍🏻 

> This syntax doesn't make sense to me.

Did my follow up post below make the syntax / expectation a bit more clear? 

Breaking down what I meant, `concat([], apply x)` was my way of saying "x is evaluated, and then concatenated against an empty array". 

In JavaScript, when you concat an array to an array, it joins them to create one array. `concat([], ['a'])` becomes `['a']`, and if you concat a value to an array, it becomes an element of the array, `concat([], 'a')` becomes `['a']`

`apply 1` just evaluates to `1`, while `apply { "+": [1,2] }` becomes `3` 


> We need to be careful about specifying this in a language/implementation-agnostic way. 

Fully agree, I was being a bit loose in my description. 

I do think that in order to support `map`, `if`, and similar operators, there does need to be some concept of being able to label operators as "lazily evaluated" to give the operator complete control of how it interprets the arguments.

When an operator is `lazy` or `not automatically traversed` or `functional` however we want to represent that, the leaf is not automatically evaluated.

**Reply by TotalTechGeek** [reply]
> My implementation doesn't perform this wrapping and instead merely operates directly on the value. I'm not sure the site [strictly speaking, it's not a spec] explicitly declares that a raw value is shorthand for a parameter array; just that the system also works with raw values. It does say "we support syntactic sugar on unary operators to skip the array around values", but that's hardly instruction to an implementation that raw values should be treated as param arrays. If this is the direction we want to take this, which could make sense, this operation needs to be more explicit in a specification.

Mine actually did not explicitly perform this either until last night. 
https://github.com/TotalTechGeek/json-logic-engine/commit/8932bcf3b77f329893f60665c00fb727370a6a8d#diff-9f6a5a68bd112cdd9fc270b72abdf7addf40ce02b6e26a036012ec0b820d2a4bL74

But this makes implementing custom operators much more consistent (so you will always receive an arguments array), and many implementations have implemented this, so I've implemented it.

**Reply by gregsdennis** [reply]
> x is evaluated, and then concatenated against an empty array

No, this doesn't make sense.  Maybe "concatenate" is the wrong word to use.  You can't concatenate non-arrays (strings are just char arrays).  You can _append_ a value to an array.

The site uses the word "cast", which I'm not sure is exactly precise, but it does make more sense than "concatenate".

> When an operator is lazy or not automatically traversed or functional however we want to represent that, the leaf is not automatically evaluated.

Even "laziness" can be an implementation decision.  We don't want to influence implementations.  We state requirements.

**Reply by TotalTechGeek** [reply]
>  Maybe "concatenate" is the wrong word to use. You can't concatenate non-arrays 

A few languages have called this operation concatenation (like Haskell), perhaps it's the wrong term to use; but I'm willing to defer on terminology; I don't really mind.

> Even "laziness" can be an implementation decision. We don't want to influence implementations. We state requirements.

OK. 

How do we want to state that operations like '+', '-', 'cat', and should expect to have this mechanism applied, 

While operations like `map`, `filter`, `reduce` are not evaluated in this way.

**Comment by TotalTechGeek** [comment]
And maybe I'll explain it in one more way (though I'll be repeating some things):

Right now, traversed arguments are passed to methods with the following rule:

`args is array ? eval args : [eval args]`

`eval x` in this is notation to execute logic. 
- `eval 1` is `1`
- `eval [1]` is `[1]`
- `eval { "+": [1, 2] }` is `3`

Imagine `{ "log": args }`

With the current rule, given data is `{"a": 1, "b": [1], "c": [1, 2] }` when args is:
- `1`, the output is `[1]`
- `"hello"`, the output is `["hello"]`
- `[1]` the output is `[1]`
- `[1,2]` the output is `[1, 2]`
- `{"var": "a"}` the output is `[1]`
- `{"var": "b"}` the output is `[[1]]`
- `[{"var": "b"}]` the output is `[[1]]`
- `{"var": "c"}` the output is `[[1, 2]]`
- `[{"var": "c"}]` the output is `[[1, 2]]`
- `{"map": [[1,2], {"+": [{"var":""}, 1]}]}` the output is `[[2,3]]`

I'd like to adjust this slightly, and I believe it would be a backwards compatible change because this is not done very often.

I'd like to switch the rule to `y = eval args, y is array ? y : [y]`, alternatively written as `concat([], eval args)`

This tweak also passes every test in the [public test suite](https://jsonlogic.com/tests.json).

With the proposed rule, given data is `{"a": 1, "b": [1], "c": [1, 2] }` when args is:
- `1`, the output is `[1]`
- `"hello"`, the output is `["hello"]`
- `[1]` the output is `[1]`
- `[1,2]` the output is `[1, 2]`
- `{"var": "a"}` the output is `[1]`
- `{"var": "b"}` the output is `[1]`
- `[{"var": "b"}]` the output is `[[1]]`
- `{"var": "c"}` the output is `[1, 2]`
- `[{"var": "c"}]` the output is `[[1, 2]]`
- `{"map": [[1,2], {"+": [{"var":""}, 1]}]}` the output is `[2,3]`

This would make it significantly easier to chain logic together, so you could call `max` or `+` or `cat` on the output of another method easily.

**Reply by gregsdennis** [reply]
> - `args is array ? eval args : [eval args]`
> - `y = eval args, y is array ? y : [y]`

These are the same.  In both cases, you're processing the `eval args` and determining if the output is an array.  If not, wrap it in an array.

I'm suggesting that we do it on _input_.  Something more akin to

```
args is array ? eval args : eval [args]

or 

y = args is array ? args : [args], eval y
```

Doing this, you can define that specific operators always take arrays of arguments, and that there's a potential "wrap" in order to adapt arguments.  Other operators might explicitly take values, and this "wrapping" doesn't occur.

**Reply by gregsdennis** [reply]
I do think we're arguing for the same thing, though.

**Reply by TotalTechGeek** [reply]
I think we're arguing for the same thing -- I think we're missing each other on the best way to represent it / how we understand it to work 😂

That first bullet point is supposed to represent: "If the argument on input is an array, eval it, or wrap it in an array and eval it"
The second bullet point represents "If the argument on output is an array, pass it along or wrap it"

I *think* we're aiming for the same outcome,

I appreciate your input.

I don't think `args is array ? eval args : eval [args]` will work, because if args returns an array, like `{ "var": "data" }`, it'll be inside of an array, *unless* we have a difference in how we interpret `eval [args]`, where it gets unwrapped from the array automatically once evaluated. Which might be because I introduced a pseudosyntax without explaining how it works / gets interpreted 😳

**Reply by TotalTechGeek** [reply]
My current conceptual understanding for JSON Logic is that there are two classifications of operator, 

- Operators that are resolved from the leaf, and MUST coerce the leaf to an array upon evaluation. Eagerly traversed. (I liked your explanation here: https://github.com/orgs/json-logic/discussions/2#discussioncomment-11477112)
- Operators that are resolved from the root, where the raw arguments are passed along to the operator. 

Those might need rephrased or massaging. I welcome your input.

Most of the operators in JSON Logic will be of the first category. `+`, `-`, `*`, `cat`, `max`, `min`, `var`. 
**Even the unary operators**, as the unary operators will only look at the first argument of the argument list they receive. So `!` and `!!` and so on.

The higher order operators like `map`, `filter`, `reduce`, `if` are responsible for traversing and handling the raw arguments, without any casting / coercion.   

I believe this convention makes it simple for folks to implement custom operators with semi-consistent types. But conventions are not ideal for specs.

I suppose we could omit the traversal piece, since that could be implementation dependent. `datalogic-rs` evaluated the arguments within the `+` operator directly. 

But I believe to ensure `{ "max": { "var: "data" } }` works as intended and maximum compatibility between operators, the array coercion must happen once the leaf is evaluated.
